tls/s2n_handshake_io.c

```diff
@@ -854,36 +860,10 @@ static int s2n_handshake_handle_sslv2(struct s2n_connection *conn)
     return 0;
 }
 
-/* Sometimes the handshake state machine can be blocked on application data. 
- * For example, if the s2n server is using a remote upstream peer for session cache lookup,
- * the handshake state machine will be blocked until there is data returned from the 
- * session cache server. In such a case, there is no data in the underlying io, 
- * so we can just call the corresponding handler to process the application data.
- * Note that we need a return value -2 to indicate that app data is not yet avaiable 
- * and the state machine is still blocked; a return value 0 to indicate the app data is 
- * successfully processed and state machine advanced; -1 to indicate an error and 
- * the connection been killed
- */
-static int s2n_handshake_handle_app_data(struct s2n_connection *conn) {
-    int r = ACTIVE_STATE(conn).handler[conn->mode] (conn);
-
-    if (r == S2N_SUCCESS) {
-        /* if r == 0, then we advance the state machine */
-        GUARD(s2n_stuffer_wipe(&conn->handshake.io));
-        GUARD(s2n_advance_message(conn));
-        goto done;
-    }
-
-    if (r < 0) {
-        if(s2n_errno != S2N_ERR_BLOCKED) {
-            GUARD(s2n_stuffer_wipe(&conn->handshake.io));
-            GUARD(s2n_connection_kill(conn));
-        }
-        return r;
-    }
-
-done:
-    conn->in_status = ENCRYPTED;
+static int s2n_try_delete_session_cache(struct s2n_connection *conn)
+{
+    notnull_check(conn);
+    conn->config->cache_delete(conn, conn->config->cache_delete_data, conn->session_id, conn->session_id_len);
```

Comment:
`notnull_check` on `conn->config` and `conn->config->cache_delete`?

Perhaps we should have a helper:

```
s2n_async_session_cache_enabled(conn) {
  notnull_check(conn);
  notnull_check(conn->config);
  return conn->config->cache_delete && conn->config->cache_store && conn->config->cache_retrieve;
}
```

Then we can guard on it?

Reply:
We have a helped, `s2n_allowed_to_cache_connection`. I've refactored the calls a little bit to place the guard inside `s2n_allowed_to_cache_connection`.